## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [65]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [66]:
# View the column names in the data
list(df.columns)
print(f"{df['category'].unique()}")




['decor' 'consumables' 'software' 'furniture' 'electronics']


In [67]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [68]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# what is the format of the phone numbers entered?
#print(f"{df['phone']}")

#check to see if the data in 'category' column is standrardized
df.count()
df = df.dropna(how="any")
df.count()
df.sum()
#print(f"{df['category']}")
#print(f"{df['category'].unique()}")







first          DonaldTiffanyShannonNathanChristinaJessicaBria...
last           HardingMyersWatsonBakerSchwartzReyesMooreWrigh...
job            ImmunologistMusic therapistImmunologistAccount...
phone          793-904-7725x39308201.442.4543x942687.737.9424...
email          harding.donald.7185@sullivan.commyers.t.6537@f...
client_id                                             2996286347
order_id                                            298885713037
order_date     2023-04-282023-05-192023-01-292023-04-252023-0...
order_week                                                620652
order_year                                             110534318
item_id        EUD29711-63-6UXDA18116-89-4AABE59463-05-7EZMM0...
category       decorconsumablessoftwareconsumablesconsumables...
subcategory    wall artpensproject managementpensmiscplantski...
unit_price                                            7445503.92
unit_cost                                             5433633.97
unit_weight              

In [69]:
# What three item categories had the most entries?
temp = dict(df.value_counts(df["category"]).nlargest(3))
print(f"{temp}")


{'consumables': 23538, 'furniture': 11915, 'software': 8400}


In [70]:
# For the category with the most entries,
# which subcategory had the most entries?
temp = dict(df.value_counts(df["subcategory"]).nlargest(1))
print(f"{temp}")


{'bathroom supplies': 6424}


In [71]:
# Which five clients had the most entries in the data?
df.value_counts(df["client_id"]).nlargest(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [72]:
#Store the client ids of those top 5 clients in a list.
top_five_clients = dict(df.value_counts(df["client_id"]).nlargest(5))
[print(f"{key}") for key in top_five_clients.keys()]



33615
66037
46820
38378
24741


[None, None, None, None, None]

In [73]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_dict = dict(df.value_counts(df["client_id"]).nlargest(1))
top_client = list(top_client_dict.keys())[0]
top_client_df = df.loc[df["client_id"] == top_client]
top_client_df['qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [74]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal$'] = df.apply(lambda _:'', axis=1)
count_row = int(df.shape[0])
for i in range(0,count_row):
    df.loc[i,'subtotal$']=df['unit_price'][i]*df['qty'][i]



In [75]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['shippingPrice'] = df.apply(lambda _:'', axis=1)
count_row = int(df.shape[0])
for i in range(0,count_row):
    order_weight = df['unit_weight'][i]*df['qty'][i]
    df.loc[i,'shippingPrice']=10*order_weight
    if (order_weight>50):
        df.loc[i,'shippingPrice']=7*order_weight
 
    
 



In [76]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['totalPrice'] = df.apply(lambda _:'', axis=1)
count_row = int(df.shape[0])
for i in range(0,count_row):
    df.loc[i,'totalPrice']=1.0925*((df['shippingPrice'][i]+df['subtotal$'][i]))



In [77]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['totalCost'] = df.apply(lambda _:'', axis=1)
count_row = int(df.shape[0])
for i in range(0,count_row):
    df.loc[i,'totalCost']=df['shippingPrice'][i]+df['unit_cost'][i]*df['qty'][i]


In [78]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df.apply(lambda _:'', axis=1)
count_row = int(df.shape[0])
for i in range(0,count_row):
    df.loc[i,'profit']= df['totalPrice'][i] - df['totalCost'][i]


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [79]:
# Check your work using the totals above
order_id_df1 = df.loc[df['order_id'] == 2742071]
print(f"Order ID 2742071 had a total price of ${order_id_df1['totalPrice'].sum():,.2f}")
order_id_df1 = df.loc[df['order_id'] == 2173913]
print(f"Order ID 2173913 had a total price of ${order_id_df1['totalPrice'].sum():,.2f}")
order_id_df1 = df.loc[df['order_id'] == 6128929]
print(f"Order ID 6128929 had a total price of ${order_id_df1['totalPrice'].sum():,.2f}")


Order ID 2742071 had a total price of $152,811.90
Order ID 2173913 had a total price of $162,388.72
Order ID 6128929 had a total price of $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [80]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
for client in top_five_clients:
   client_df = df.loc[df['client_id'] == client, 'totalPrice']
   print(f"{client}: ${client_df.sum():,.2f}")
   
  

33615: $8,377,308.57
66037: $10,259,514.80
46820: $9,743,794.32
38378: $12,906,550.88
24741: $82,268,891.98


In [81]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

#top_five_clients
client_df1 = pd.DataFrame()
client_df1['client'] = client_df1.apply(lambda _:0, axis=1)
client_df1['total units'] = client_df1.apply(lambda _:0, axis=1)
client_df1['total shipping price'] = client_df1.apply(lambda _:0, axis=1)
client_df1['total revenue'] = client_df1.apply(lambda _:0, axis=1)
client_df1['total cost'] = client_df1.apply(lambda _:0,axis=1)
client_df1['total profit'] = client_df1.apply(lambda _:0, axis=1)

x=0
for client in top_five_clients:
    client_df1.loc[x,"client"] = int(client)
    client_df1.loc[x,"total units"]= df.loc[df['client_id'] == client, 'qty'].sum()
    client_df1.loc[x,"total shipping price"]= df.loc[df['client_id'] == client, 'shippingPrice'].sum()
    client_df1.loc[x,"total revenue"]= df.loc[df['client_id'] == client, 'totalPrice'].sum()
    client_df1.loc[x,'total cost'] = df.loc[df['client_id'] == client, 'totalCost'].sum()
    client_df1.loc[x,"total profit"]= df.loc[df['client_id'] == client, 'profit'].sum()
    x+=1


client_df1 = client_df1.sort_values(by="total profit", ascending=False)
print(f"{client_df1}")

    client  total units  total shipping price  total revenue   total cost  \
4  24741.0     239862.0            5126448.37   8.226889e+07  45688899.71   
3  38378.0      73667.0            3429455.40   1.290655e+07   9634720.98   
1  66037.0      43018.0            1395151.85   1.025951e+07   7004482.98   
2  46820.0      75768.0            1601448.84   9.743794e+06   7007191.64   
0  33615.0      64313.0            1828984.89   8.377309e+06   6175313.91   

   total profit  
4  3.657999e+07  
3  3.271830e+06  
1  3.255032e+06  
2  2.736603e+06  
0  2.201995e+06  


In [82]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

x=0
count = len(client_df1)
for x in range(0,count):
    client_df1.loc[x,"total units"]= round(client_df1.loc[x,"total units"],0)
    client_df1.loc[x,"total shipping price"]= round(client_df1.loc[x,"total shipping price"]/1000000,2)
    client_df1.loc[x,"total revenue"]= round(client_df1.loc[x,"total revenue"]/1000000,2)
    client_df1.loc[x,"total cost"]= round(client_df1.loc[x,"total cost"]/1000000,2)
    client_df1.loc[x,"total profit"] = round(client_df1.loc[x,"total profit"]/1000000,2)
    x+=1


client_df1 = client_df1.rename(columns ={"client":"ClientID", 
                                         "total units":"Units", 
                                         "total shipping price":"Shipping$",
                                         "total revenue":"Revenue$",
                                         "total cost" : "Cost$",
                                         "total profit" :"Profit$"})



print(f"{client_df1}")



   ClientID     Units  Shipping$  Revenue$  Cost$  Profit$
4   24741.0  239862.0       5.13     82.27  45.69    36.58
3   38378.0   73667.0       3.43     12.91   9.63     3.27
1   66037.0   43018.0       1.40     10.26   7.00     3.26
2   46820.0   75768.0       1.60      9.74   7.01     2.74
0   33615.0   64313.0       1.83      8.38   6.18     2.20


In [84]:
# Sort the updated data by "Total Profit" form highest to lowest
sorted_client_df = client_df1.sort_values(by="Profit$", ascending=False)
columns = list(sorted_client_df)
count = int(sorted_client_df.shape[0])-1

print(*columns)
for i in range(count,-1,-1):
    print(f"{sorted_client_df.loc[i,'ClientID']:.0f}\t{sorted_client_df.loc[i,'Units']:,.0f}\t${sorted_client_df.loc[i,'Shipping$']:.2f}\t${sorted_client_df.loc[i,'Revenue$']:.2f}\t ${sorted_client_df.loc[i,'Cost$']:.2f}\t${sorted_client_df.loc[i,'Profit$']:.2f}")
    


ClientID Units Shipping$ Revenue$ Cost$ Profit$
24741	239,862	$5.13	$82.27	 $45.69	$36.58
38378	73,667	$3.43	$12.91	 $9.63	$3.27
46820	75,768	$1.60	$9.74	 $7.01	$2.74
66037	43,018	$1.40	$10.26	 $7.00	$3.26
33615	64,313	$1.83	$8.38	 $6.18	$2.20
